### PR TITLE
Extend source classification to all kinds of GraceDB uploads

### PR DESCRIPTION
**Pull request by veronica-villa**
Currently the source classification in PyCBC Live is only performed for triggers resulting from a two-detector coincidence. This PR extends it for the triggers resulting from the SNR maximization followup (`pycbc_optimize_snr`) as well as single-detector triggers (`check_singles()` in `pycbc_live`).

### CODE REVIEW

**Comment by veronica-villa** [comment]
An example of a trigger from SNR optimization can be seen in: https://gracedb-playground.ligo.org/events/T256619/view/

File: bin/pycbc_optimize_snr [hunk 1]
```diff
@@ -319,7 +320,9 @@ for ifo in fp['channel_names'].keys():
 kwargs = {'psds': {ifo: followup_data[ifo]['psd'] for ifo in ifos},
           'low_frequency_cutoff': flow,
           'followup_data': followup_data,
-          'channel_names': channel_names}
+          'channel_names': channel_names,
+          'mc_area_args': recursively_load_hdf5_group_to_dict(fp,
```
Comment (titodalcanton) [inline bin/pycbc_optimize_snr]: Just a minor code readability point: here I would do
```python
mc_area_args = recursively_load_hdf5_group_to_dict(fp, ...)
kwargs = {'psd': ...
          'mc_area_args': mc_area_args}
```

File: pycbc/io/hdf.py [hunk 2]
```diff
@@ -960,13 +960,53 @@ def recursively_save_dict_contents_to_group(h5file, path, dic):
...
-        if isinstance(item, (np.ndarray, np.int64, np.float64, str, bytes, tuple, list)):
+        if isinstance(item, (np.ndarray, np.int64, np.float64, str, int, float,
+                             bytes, tuple, list)):
             h5file[path + str(key)] = item
         elif isinstance(item, dict):
             recursively_save_dict_contents_to_group(h5file, path + key + '/', item)
         else:
             raise ValueError('Cannot save %s type' % type(item))
 
+def load_hdf5_to_dict(filename):
```
Comment (titodalcanton) [inline pycbc/io/hdf.py]: Do you actually use this function? If not, please remove.

File: pycbc/io/hdf.py [hunk 3]
```diff
@@ -960,13 +960,53 @@ def recursively_save_dict_contents_to_group(h5file, path, dic):
...
+    filename:
+	hdf5 file to be loaded as a dictionary
+
+    Returns
+    -------
+    dictionary with hdf5 file content
+
+    """
+    with h5py.File(filename, 'r') as h5file:
+        recursively_load_hdf5_group_to_dict(h5file, '/')
```
Comment (titodalcanton) [inline pycbc/io/hdf.py]: Looks like a `return` is missing here (but see previous comment).

File: pycbc/io/hdf.py [hunk 4]
```diff
@@ -960,13 +960,53 @@ def recursively_save_dict_contents_to_group(h5file, path, dic):
...
+
+def recursively_load_hdf5_group_to_dict(h5file, path):
```
Comment (titodalcanton) [inline pycbc/io/hdf.py]: Can this function be called simply `load_hdf5_to_dict()`?

File: pycbc/mchirp_area.py [hunk 5]
```diff
@@ -320,7 +320,7 @@ def calc_probabilities(mchirp, snr, eff_distance, src_args):
     # equal to the maximum mass, the probability for BBH is 100%
     mc_max = mass_limits['max_m1'] / (2 ** 0.2)
     if trig_mc_det['central'] > mc_max * (1 + z['central']):
-        if mass_gap is not False:
+        if mass_gap:
```
Comment (titodalcanton) [inline pycbc/mchirp_area.py]: 👍

**Comment by titodalcanton** [comment]
Thanks @veronica-villa, this looks great. Please see the comments I left and the code climate issues.

File: pycbc/io/hdf.py (hunk 2 above)
Comment (veronica-villa) [reply pycbc/io/hdf.py]: I wrote it this way to mimic the two already existing functions `recursively_save_dict_contents_to_group` and `save_dict_to_hdf5`,  but it's simpler as you say

File: pycbc/io/hdf.py [hunk 6]
```diff
@@ -960,13 +960,38 @@ def recursively_save_dict_contents_to_group(h5file, path, dic):
...
+    Returns
+    -------
+    dic:
+        dictionary with hdf5 file group content
+    """
+    dic = {}
+    for key, item in h5file[path].items():
+        if isinstance(item, h5py._hl.dataset.Dataset):
+            dic[key] = item[()]
+        elif isinstance(item, h5py._hl.group.Group):
```
Comment (titodalcanton) [inline pycbc/io/hdf.py]: This is raising some more Codeclimate issues. Is it possible to just use `h5py.Group` and `h5py.Dataset` here? Or is there a specific reason to use those in `_hl.dataset`?

Comment (veronica-villa) [reply pycbc/io/hdf.py]: I just used the output I got from doing `type(h5file[path])` for groups and datasets but I checked and `h5py.Group` and `h5py.Dataset` are valid types too, so thanks for solving this

Comment (titodalcanton) [reply pycbc/io/hdf.py]: Ok nice, but make sure it works! They may be slightly different derived classes.

Comment (veronica-villa) [reply pycbc/io/hdf.py]: I've checked it with ipython and it seems to work correctly

**Comment by titodalcanton** [comment]
The Travis failure is unrelated to this, so merging.

**Pull request merged**